### PR TITLE
Add ON CUDA kernel

### PR DESCRIPTION
**Pull request by vchiley**
Hey Atli,

This PR adds the ON kernel to the PyTorch code (Vishal is working on integrating the kernels with TF).
You've written kernels before so it looks like you get to review this. lol. Also, I don't know who else on the ML team has written a CUDA kernel before.

Notes:
1. I'm leaving the BON code there and throwing a Deprecated warning. Reasoning: all of our Experiments were done with that implementation.
1. The CUDA kernels were written in an effort to absorb the loops associated with ON. Most of the reductions required are done with ATen member methods (add, sub, multi, div, mean, etc.). This is done for the fwd process and the bwd u controller. The v controller has a reduction in the loop so I had to do that myself.
1. Without knowledge of the ON algorithm, Philippe looked at the kernels and thinks they look good.

### CODE REVIEW

File: online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+      s_u_local = s_u_local + (1. - abkw) * delta;
+    };
+
+    s_u[c] = s_u_local;
+  };
+}
+
+__device__ void warp_reduce(volatile float *s_mem, const unsigned int t_id, const unsigned int d) {
+  for (unsigned int ridx = 32; ridx > 0; ridx /= 2) {
+    if (d > ridx) { if (t_id < ridx) { if ((t_id + ridx) < d) { s_mem[t_id] += s_mem[t_id + ridx]; } } }
```
Comment (atlikosson) [inline online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu]: Might need a __syncwarp(), probably OK but I'm not 100% sure. Would likely be cleaner to write this out over multiple lines and potentially use &&.

File: online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    __syncthreads();
+
+    // reduce within thread block % warp reduction
+    for (idx = 512; idx > 32; idx /= 2) {
+      if (d > idx) { if (t_id < idx) { if ((t_id + idx) < d) { s_mem[t_id] += s_mem[t_id + idx]; } } __syncthreads(); }
+    }
+    if (t_id < 32) { warp_reduce(s_mem, t_id, d); }     // reduce within warp
+
+    // update vctrl / mv reduction to global mem
+    if (t_id == 0) { s_v[c] += (s_mem[0] / D); };
```
Comment (atlikosson) [inline online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu]: extra ;, also after for {} block

File: online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+  unsigned int idx3, idx;
+
+  extern __shared__ float s_mem[];
+
+  for(int n = 0; n < N; ++n){
+    s_mem[t_id] = 0;                                // reset shared mem
+    for (idx = t_id; idx < D; idx += d) {
+      idx3 = Idx3(n, c, idx, N, C, D);              // idx in global mem
+      // vctrl logic
+      grad_tmp[idx3] = grad_out[idx3] - (1. - abkw) * (scalar_t)(s_v[c]) * out[idx3];
```
Comment (atlikosson) [inline online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu]: I'm a little confused about the datatypes. I'm assuming scalar_t can be __half or float. Does the presence of abkw cast s_v[c] back to float?

Is the reason you want to support __half inside the boundaries of the layer at all performance / bandwidth saving or to try to match some other reference? It seems to require a bunch of casting.

**Review by atlikosson — APPROVED** [review]
Looks good to me

File: online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu (hunk 3 above)
Comment (vchiley) [reply online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu]: you're right `abkw` should be cast to `scalar_t`

File: online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu (hunk 2 above)
Comment (vchiley) [reply online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu]: true

**Pull request merged**